internal/mining/tx_desc_graph.go

```diff
@@ -0,0 +1,246 @@
+	seen map[chainhash.Hash]struct{}, f func(tx *TxDesc)) {
+
+	for parent, parentDesc := range g.parentsOf[*txHash] {
+		if _, saw := seen[parent]; saw {
+			continue
+		}
+
+		seen[parent] = struct{}{}
+		g.forEachAncestor(&parent, seen, f)
+		f(parentDesc)
+	}
+}
+
+// forEachAncestorPreOrder iterates over all transactions in the graph that
+// txHash depends on and invokes the function f for each, in pre-order.
+// If the provided function f returns true then it continues to walk ancestors
+// of the respective ancestor. If it returns false, then no additional parents
+// of the provided transaction will be visited and the transaction passed to f
+// will not be added to the seen map.
+func (g *txDescGraph) forEachAncestorPreOrder(txHash *chainhash.Hash,
+	seen map[chainhash.Hash]*TxDesc, f func(tx *TxDesc) bool) {
+
+	for parentHash, parentTxDesc := range g.parentsOf[*txHash] {
+		if _, saw := seen[parentHash]; saw {
+			continue
+		}
+
+		moveNext := f(parentTxDesc)
+		if !moveNext {
+			return
+		}
+
+		seen[parentHash] = parentTxDesc
+		g.forEachAncestorPreOrder(&parentHash, seen, f)
+	}
+}
+
+// forEachDescendant iterates depth-first over all transactions that depend on
+// the provided transaction hash and invokes function f with each in post-order.
+func (g *txDescGraph) forEachDescendant(txHash *chainhash.Hash,
+	seen map[chainhash.Hash]struct{}, f func(*TxDesc)) {
+
+	for child, childDesc := range g.childrenOf[*txHash] {
+		if _, saw := seen[child]; saw {
+			continue
+		}
+
+		seen[child] = struct{}{}
+		g.forEachDescendant(&child, seen, f)
+		f(childDesc)
+	}
+}
+
+// forEachDescendantPreOrder attempts to walk all transactions that depend on
+// the provided transaction hash by invoking the function f with each in
+// pre-order. If the provided function f returns true then the traversal will
+// walk descendants of the provided transaction's respective child.
+func (g *txDescGraph) forEachDescendantPreOrder(txHash *chainhash.Hash,
+	seen map[chainhash.Hash]struct{}, f func(*TxDesc) bool) {
+
+	for child, childDesc := range g.childrenOf[*txHash] {
+		if _, saw := seen[child]; saw {
+			continue
+		}
+
+		seen[child] = struct{}{}
+		if f(childDesc) {
+			g.forEachDescendantPreOrder(&child, seen, f)
+		}
+	}
+}
+
+// When a transaction is added to the mempool, create a 2-way association
+// between itself and its parents in the mempool.
+func (g *txDescGraph) insert(txDesc *TxDesc, findTx txDescFind) {
+	seen := make(map[chainhash.Hash]struct{})
+
+	// Fetch transactions that spend this one from the graph.
+	g.forEachRedeemer(txDesc.Tx, func(child *TxDesc) {
+		g.addChild(txDesc, child)
+		g.addParent(child, txDesc)
+	})
+
+	// Relate self with direct ancestors.
+	for _, txIn := range txDesc.Tx.MsgTx().TxIn {
+		parentHash := txIn.PreviousOutPoint.Hash
+		if _, saw := seen[parentHash]; saw {
+			continue
+		}
+
+		seen[parentHash] = struct{}{}
+
+		// Find parents from the mempool and add to graph.
```

Comment:
This can no longer definitively say where it is finding them.  Perhaps: `Find parents using the provided locator and add to graph.`?